### PR TITLE
[PM-8829] Password/Fido2 credentials autofill on prepare credential list

### PR DESCRIPTION
**Pull request by fedemkr**
## 🎟️ Tracking

[PM-8829](https://bitwarden.atlassian.net/browse/PM-8829)

## 📔 Objective

Handle prepare credential list for autofilling passwords/Fido2 credentials from the iOS Autofill extension.

## 📸 Screenshots

https://github.com/user-attachments/assets/7bfdabcc-6d3b-4cff-951b-f7c293aeef3c

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

[PM-8829]: https://bitwarden.atlassian.net/browse/PM-8829?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: BitwardenShared/UI/Vault/Vault/AutofillList/VaultAutofillListProcessor.swift [hunk 1]
```diff
@@ -354,21 +490,68 @@ extension VaultAutofillListProcessor {
...
+        let fido2ListItems = try await filteredFido2Credentials
+            .asyncMap { cipher in
+                let fido2CredentialAutofillView = try await self.services.clientService
+                    .platform()
+                    .fido2()
+                    .decryptFido2AutofillCredentials(cipherView: cipher)
+
+                return VaultListItem(
+                    cipherView: cipher,
+                    fido2CredentialAutofillView: fido2CredentialAutofillView[0]
```
Comment (matt-livefront) [inline BitwardenShared/UI/Vault/Vault/AutofillList/VaultAutofillListProcessor.swift:524]: ❓ Is there any risk that `fido2CredentialAutofillView` would be an empty array? Any reason to not add a guard for getting `fido2CredentialAutofillView.first` instead?

File: BitwardenShared/UI/Vault/Vault/AutofillList/VaultAutofillListView.swift [hunk 2]
```diff
@@ -107,7 +107,35 @@ private struct VaultAutofillListSearchableView: View {
 
     /// A view for displaying a list of ciphers.
     @ViewBuilder
-    private func cipherListView(_ items: [VaultListItem]) -> some View {
+    private func cipherListView(_ sections: [VaultListSection]) -> some View {
+        if store.state.isAutofillingFido2List {
+            cipherCombinedListView(sections)
+        } else {
+            let items = sections[safeIndex: 0]?.items ?? []
```
Comment (matt-livefront) [inline BitwardenShared/UI/Vault/Vault/AutofillList/VaultAutofillListView.swift:114]: 🎨 I think I'd prefer using `sections.first` over the subscript here.
```suggestion
            let items = sections.first?.items ?? []
```

File: BitwardenShared/Core/Autofill/Models/PasskeyCredentialRequestParameters.swift [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+public protocol PasskeyCredentialRequestParameters {
+    /// The relying party identifier for this request.
+    var relyingPartyIdentifier: String { get }
+    /// Hash of client data for credential provider to sign as part of the operation.
+    var clientDataHash: Data { get }
+    /// A preference for whether the authenticator should attempt to verify that it is being used by its owner,
+    /// such as through a PIN or biometrics.
+    var userVerificationPreference: ASAuthorizationPublicKeyCredentialUserVerificationPreference { get }
+    /// A list of allowed credential IDs for this request. An empty list means all credentials are allowed.
+    var allowedCredentials: [Data] { get }
```
[the remark this answers is not in the document]
Comment (fedemkr) [reply BitwardenShared/Core/Autofill/Models/PasskeyCredentialRequestParameters.swift:14]: Oh copy pasted Apple's properties hehe, I'll change it, thanks for spotting it.

File: BitwardenShared/UI/Vault/Vault/AutofillList/VaultAutofillListProcessor.swift [hunk 4]
```diff
@@ -290,16 +352,90 @@ extension VaultAutofillListProcessor {
...
+                clientDataHash: fido2RequestParameters.clientDataHash,
+                authenticatorData: assertionResult.authenticatorData,
+                credentialID: assertionResult.credentialId
+            ))
+        } catch {
+            #if DEBUG
+            Fido2DebuggingReportBuilder.builder.withGetAssertionResult(.failure(error))
+            #endif
+            services.fido2UserInterfaceHelper.pickedCredentialForAuthentication(result: .failure(error))
+            services.errorReporter.log(error: error)
```
[the remark this answers is not in the document]
Comment (fedemkr) [reply BitwardenShared/UI/Vault/Vault/AutofillList/VaultAutofillListProcessor.swift:438]: I can try that to have a cleaner code.

**Review by matt-livefront — APPROVED** [review]

File: BitwardenShared/Core/Vault/Repositories/VaultRepositoryTests.swift [hunk 5]
```diff
@@ -137,24 +140,115 @@ class VaultRepositoryTests: BitwardenTestCase { // swiftlint:disable:this type_b
...
+            uri: "https://example.com"
+        ).makeAsyncIterator()
+        let sectionsResult = try await iterator.next()
+        let sections = try XCTUnwrap(sectionsResult)
+
+        XCTAssertEqual(
+            sections[0],
+            VaultListSection(
+                id: Localizations.passkeysForX(expectedRpID),
+                items: expectedCiphersInFido2Section.map { cipherView in
```
[the remark this answers is not in the document]
Comment (fedemkr) [reply BitwardenShared/Core/Vault/Repositories/VaultRepositoryTests.swift:219]: Oh you're right, I totally missed inline snapshots! 🎉 I'll make a tech-debt task for this so I can tackle it properly on another PR given that I'd need some update in the dump so it shows something about the inner Fido2 credentials in order to see that the item section is complete with all the information it needs.

**Pull request merged**